src/OrchardCore/OrchardCore.Environment.Cache/TenantServicesBuilderExtensions.cs

```diff
@@ -26,10 +26,7 @@ public static TenantServicesBuilder AddCaching(this TenantServicesBuilder tenant
             services.AddScoped<ICacheContextProvider, UserCacheContextProvider>();
             services.AddScoped<ICacheContextProvider, KnownValueCacheContextProvider>();
 
-            // MVC is already registering IMemoryCache as host singleton. We are registering it again
-            // in this module so that there is one instance for each tenant.
-            // Important: we can't call AddMemoryCache as it's using the TryAdd pattern and hence would
-            // not override any existing instance defined at the host level by MVC
+            // IMemoryCache is registered at the tenant level so that there is one instance for each tenant.
             services.AddSingleton<IMemoryCache, MemoryCache>();
 
             // MemoryDistributedCache needs to be registered as a singleton as it owns a MemoryCache instance.
```

Comment:
I was actually thinking about this and here it is already done. 👍

Reply:
@Skrypt 

About `IMemoryCache`, it is registered by an mvc helper which is called when we call our `AddMvcModules()` (renamed here). In this case, `IMemoryCache` is already registered as a per tenant singleton (not an app singleton). But there are multiple scenarios, e.g:

1. An Orchard mvc app which call our helper => nothing to do.

2. A non mvc Orchard app but want to use our tenant level caching => we need to register it as we do and here we could use `AddMemoryCache()`.

3. Same as 2. but the app also want to use the regular `.AddCaching()` but at the app level which here would add an application singleton that would be also shared by all tenants => we need to register it as we do and here we can't use `AddMemoryCache()` with its `.TryAdd()` pattern.

About `MemoryDistributedCache` it is registered e.g by the mvc helper `AddCacheTagHelper()` but we we don't call it. So, you're right it is not registered in the first above scenario and we have to do it. Then, for the same reason as above we can't use `AddDistributedMemoryCache()`.

Note: Yes, we don't use the aspnet `AddMvc()`, we only call `AddMvcCore()` with the essential mvc services because we need to compose then with mvc and custom orchard services. Hmm, this let me think that we need to revisit what the regular `AddMvc()` now do to be sure we don't loose anything useful, e.g we don't call the regular `AddCacheTagHelper()`, `AddRazorPages()`.

Then, as i remember, we also register as a tenant singleton a `DefaultDynamicCache` which is only a wrapper so, as i understand, it adds nothing in term of implementation but it is registered by the `OC.DynamicCache` so that it is available only if the related feature is enabled and then it will use the currently registered `IDistributedCache`.

Reply:
I think you got this under control 😉